src/solver.rs

```diff
@@ -124,79 +128,215 @@ pub fn resolve<P: Package, V: Version>(
-                            });
-                        }
-                        x
-                    }
-                };
-
-            // Add that package and version if the dependencies are not problematic.
-            let dep_incompats =
-                state.add_incompatibility_from_dependencies(p.clone(), v.clone(), &dependencies);
-
-            // TODO: I don't think this check can actually happen.
-            // We might want to put it under #[cfg(debug_assertions)].
-            if state.incompatibility_store[dep_incompats.clone()]
-                .iter()
-                .any(|incompat| state.is_terminal(incompat))
-            {
-                // For a dependency incompatibility to be terminal,
-                // it can only mean that root depend on not root?
-                return Err(PubGrubError::Failure(
-                    "Root package depends on itself at a different version?".into(),
-                ));
+                        dependent: dependent.clone(),
+                    });
+                }
+
+                x
             }
-            state.partial_solution.add_version(
-                p.clone(),
-                v,
-                dep_incompats,
-                &state.incompatibility_store,
-            );
-        } else {
+        };
+
+        // Add that package and version if the dependencies are not problematic.
+        let dep_incompats =
+            state.add_incompatibility_from_dependencies(p.clone(), v.clone(), &dependencies);
+
+        // TODO: I don't think this check can actually happen.
+        // We might want to put it under #[cfg(debug_assertions)].
+        if state.incompatibility_store[dep_incompats.clone()]
+            .iter()
+            .any(|incompat| state.is_terminal(incompat))
+        {
+            // For a dependency incompatibility to be terminal,
+            // it can only mean that root depend on not root?
+            return Err(PubGrubError::Failure(
+                "Root package depends on itself at a different version?".into(),
+            ));
+        }
+
+        state.partial_solution.add_version(
+            p.clone(),
+            v,
+            dep_incompats,
+            &state.incompatibility_store,
+        );
+    }
+}
+
+/// Main function of the library.
+/// Finds a set of packages satisfying dependency bounds for a given package + version pair.
+#[cfg(feature = "async")]
+pub async fn resolve_async<P: Package, V: Version>(
+    dependency_provider: &impl AsyncDependencyProvider<P, V>,
+    package: P,
+    version: impl Into<V>,
+) -> Result<SelectedDependencies<P, V>, PubGrubError<P, V>> {
+    let mut state = State::init(package.clone(), version.into());
+    let mut added_dependencies: Map<P, Set<V>> = Map::default();
+    let mut next = package;
+    loop {
+        dependency_provider
+            .should_cancel()
+            .map_err(|err| PubGrubError::ErrorInShouldCancel(err))?;
+
+        state.unit_propagation(next)?;
```

Comment:
I was very lucky to get @carllerche's opinion on this. The two alternatives he suggested were:

1. Have `identify_problems` to a fixed amount of work and if there's more work to be done do the rest in a `spon_blocking`. In theory, this is the best solution. In practice I'm not sure that we can clearly identify how much work is being done by each sub call within `unit_propagation`, so I don't know if it's practical. We would also need extensive tests to verify that we don't accidentally have a loop occurring before we call `spon_blocking`.
2. Do not provide an async interface. Callers can `block_in_place` before calling resolve, and call `tokio::runtime::Handle::block_on` within the callbacks.

I'm not sure which solution is more unpleasant/inefficient, but it's comforting to know that it is a real problem.

Reply:
To be honest, I did not know of the `block_in_place` approach, but that sounds like something that might be easy to use and still provide the ability to interface with async code.

Maybe one idea is to add this to the documentation with some code examples? This should be a good starting point for other people that interface with async stuff.

Reply:
My main concern with that approach is how it works with other executors. It can work well for Tokyo, but are there equivalents in the others? What about when running as Wasm in a browser?

Anyway, 110% on documentation. We should have a section in the book about how to shoehorn asynchronous communication into the current API. That, to the best of our ability, documents the available options and their advantages and disadvantages. That unblocks people without waiting for us to release a new version of the crate.